File: src/aoc2022/dijkstra.clj
(ns aoc2022.dijkstra
  (:require [aoc2022.pq :refer :all]))

(defn make-path
  [prev target]
  (loop [n target
         path []]
    (if (nil? n)
      (reverse path)
      (recur (prev n) (conj path n))
      )
    )
  )

(defn shortest-path
  [start end adj-fn cost-fn can-move-fn]
  (loop [pq (-> (make-pq)
                (pq-put! start 0))
         dist-to {start 0}
         prev {start nil}]
    (if (pq-empty? pq)
      (make-path dist-to end)
      (let [[current pq] (pq-poll! pq)]
        (if (= current end)
          (make-path prev end)
          (let [[dist-to prev pq]
                (loop [neighbours (adj-fn current)
                       dist-to dist-to
                       prev prev
                       pq pq]
                  (if (empty? neighbours)
                    [dist-to prev pq]
                    (let [next (first neighbours)
                          existing-next-distance (get dist-to next (Integer/MAX_VALUE))
                          existing-current-distance (dist-to current)
                          new-next-distance (+ existing-current-distance (cost-fn current next))]
                      (if (and (or (not (contains? dist-to next)) (< new-next-distance existing-next-distance)) (can-move-fn current next))
                        (let [dist-to (assoc dist-to next new-next-distance)
                              pq (pq-put! pq next new-next-distance)
                              prev (assoc prev next current)]
                          (recur (rest neighbours) dist-to prev pq)
                          )
                        (recur (rest neighbours) dist-to prev pq)
                        )
                      ))
                  )]
            (recur pq dist-to prev))
          )
        )
      )
    ))

File: src/aoc2022/common/Dijkstra.java
package aoc2022.common;

import java.util.*;
import java.util.function.Function;

public class Dijkstra<T> {

    private final Function<T, List<T>> neighbours;

    public Dijkstra(Function<T, List<T>> neighbours) {
        this.neighbours = neighbours;
    }

    public List<T> path(T source, T target) {
        Map<T, Integer> distTo = new HashMap<>();
        Map<T, T> prev = new HashMap<>();
        distTo.put(source, 0);
        Queue<T> q = new PriorityQueue<>(Comparator.comparingInt(distTo::get));
        q.add(source);
        prev.put(source, null);

        while (!q.isEmpty()) {
            T current = q.poll();
            if (current == target) {
                break;
            }
            for (T next : neighbours.apply(current)) {
                int dc = distTo.getOrDefault(current, Integer.MAX_VALUE - 1);
                int dn = dc + 1;
                if (!distTo.containsKey(next) || dn < distTo.getOrDefault(next, Integer.MAX_VALUE - 1)) {
                    distTo.put(next, dn);
                    prev.put(next, current);
                    q.remove(next);
                    q.add(next);
                }
            }
        }

        List<T> path = new ArrayList<>();
        T p = prev.get(target);
        while (p != null) {
            path.add(p);
            p = prev.get(p);
        }
        Collections.reverse(path);
//        System.err.println(distTo.get(target));
        return path;
    }
}
